### ISSUE TITLE
Unable to fix pgpool adminPassword

### ISSUE THREAD

**Issue by MswPaulDSmith** [bitnami/charts#8093]
**Which chart**:
https://github.com/bitnami/charts/tree/master/bitnami/postgresql-ha

**Describe the bug**
As per your article...
https://docs.bitnami.com/general/how-to/troubleshoot-helm-chart-issues/
...I am trying to fix passwords on install but cannot see how to set the pgpool.adminPassword from a kubernetes secret store.  Can you please clarify how to achieve this?  I have managed to set other password fields in this manner but the chart implies there is no way to make this field be read from the secrets store and this seems to be my experience.

**To Reproduce**

1. Try to configure use of my own secret store via values.yaml
2. Install my product
3. Confirm using "kubectl edit secrets app-postgresql-ha-pgpool" that a random password has still been created.

**Expected behavior**
Your article implies I should be able to configure my own password - how?.

**Version of Helm and Kubernetes**:

- Output of `helm version`:

```
version.BuildInfo{Version:"v3.7.1", GitCommit:"1d11fcb5d3f3bf00dbe6fe31b8412839a96b3dc4", GitTreeState:"clean", GoVersion:"go1.16.9"}
```

- Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.3", GitCommit:"c92036820499fedefec0f847e2054d824aea6cd1", GitTreeState:"clean", BuildDate:"2021-10-27T18:41:28Z", GoVersion:"go1.16.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.7", GitCommit:"1dd5338295409edcfff11505e7bb246f0d325d15", GitTreeState:"clean", BuildDate:"2021-01-13T13:15:20Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
WARNING: version difference between client (1.22) and server (1.19) exceeds the supported minor version skew of +/-1
```

**Additional context**
Really not sure if this is my failing to understand your chart documentation, a bug in the chart or the referenced article giving a false sense of what can be achieved.  Thanks for putting me right on this!

**Comment by pablogalegoc** [comment]
Hi @MswPaulDSmith!

If I understand correctly you want to use an existing secret from a secret store that you have set up.
If that's the case you can use:
https://github.com/bitnami/charts/blob/7f36a0c274817c69377f7c22bcfa68536ed41d2d/bitnami/postgresql-ha/values.yaml#L614-L617

**Comment by MswPaulDSmith** [comment]
Does this also apply to the pgpool.adminPassword then and if so, how to I indicate which of the list of users/passwords is the admin?

**Comment by pablogalegoc** [comment]
Ah, I think this is completely our fault, see that the secret creation depends on the execution of this macro:
https://github.com/bitnami/charts/blob/df4fe38f535992798554218847014d5b41f68837/bitnami/postgresql-ha/templates/pgpool/secrets.yaml#L1
Which looks if there's an existing secret referenced:
https://github.com/bitnami/charts/blob/df4fe38f535992798554218847014d5b41f68837/bitnami/postgresql-ha/templates/_helpers.tpl#L375-L388
`global.pgpool.existingSecret` is declared in the `values.yaml` but `pgpool.existingSecret` it's not. You should be able to add the latter to the values and reference a secret that holds the `admin-password` key.

If it works and you are willing to submit a PR to add `pgpool.existingSecret` to the values it would be awesome 🙂

**Comment by MswPaulDSmith** [comment]
Pablo,

Am I correct that the error is just one of documentation and you want me to update the Markdown that describes the currently undocumented setting?

If that’s correct then I have something ready to go but it doesn’t appear to have solved my problem – are you able to tell me if this is a follow-on or a new issue?

My postgresql/repmgr system is no longer complaining about passwords but still doesn’t come up.  The main issue seems to be this from what I assume is the back-up node…

***@***.*** ~$ kubectl logs hapestfvmon-postgresql-ha-postgresql-1
postgresql-repmgr 10:20:55.08
postgresql-repmgr 10:20:55.09 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 10:20:55.09 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql-repmgr
postgresql-repmgr 10:20:55.09 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql-repmgr/issues
postgresql-repmgr 10:20:55.09
postgresql-repmgr 10:20:55.12 INFO  ==> ** Starting PostgreSQL with Replication Manager setup **
postgresql-repmgr 10:20:55.15 INFO  ==> Validating settings in REPMGR_* env vars...
postgresql-repmgr 10:20:55.16 INFO  ==> Validating settings in POSTGRESQL_* env vars..
postgresql-repmgr 10:20:55.16 INFO  ==> Querying all partner nodes for common upstream node...
postgresql-repmgr 10:21:05.23 INFO  ==> Preparing PostgreSQL configuration...
postgresql-repmgr 10:21:05.24 INFO  ==> postgresql.conf file not detected. Generating it...
postgresql-repmgr 10:21:05.38 INFO  ==> Preparing repmgr configuration...
postgresql-repmgr 10:21:05.40 INFO  ==> Initializing Repmgr...
postgresql-repmgr 10:21:05.41 INFO  ==> Waiting for primary node...

The primary node appears to be working and these were the logs:

postgresql-repmgr 09:35:49.94
postgresql-repmgr 09:35:49.94 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 09:35:49.94 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql-repmgr
postgresql-repmgr 09:35:49.95 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql-repmgr/issues
postgresql-repmgr 09:35:49.95
postgresql-repmgr 09:35:50.00 INFO  ==> ** Starting PostgreSQL with Replication Manager setup **
postgresql-repmgr 09:35:50.04 INFO  ==> Validating settings in REPMGR_* env vars...
postgresql-repmgr 09:35:50.04 INFO  ==> Validating settings in POSTGRESQL_* env vars..
postgresql-repmgr 09:35:50.05 INFO  ==> Querying all partner nodes for common upstream node...
postgresql-repmgr 09:35:50.25 INFO  ==> There are no nodes with primary role. Assuming the primary role...
postgresql-repmgr 09:35:50.26 INFO  ==> Preparing PostgreSQL configuration...
postgresql-repmgr 09:35:50.28 INFO  ==> postgresql.conf file not detected. Generating it...
postgresql-repmgr 09:35:50.44 INFO  ==> Preparing repmgr configuration...
postgresql-repmgr 09:35:50.45 INFO  ==> Initializing Repmgr...
postgresql-repmgr 09:35:50.47 INFO  ==> Initializing PostgreSQL database...
postgresql-repmgr 09:35:50.48 INFO  ==> Custom configuration /opt/bitnami/postgresql/conf/postgresql.conf detected
postgresql-repmgr 09:35:50.48 INFO  ==> Custom configuration /opt/bitnami/postgresql/conf/pg_hba.conf detected
postgresql-repmgr 09:35:52.22 INFO  ==> Starting PostgreSQL in background...
postgresql-repmgr 09:35:52.36 INFO  ==> Changing password of postgres
postgresql-repmgr 09:35:52.38 INFO  ==> Creating replication user repmgr
postgresql-repmgr 09:35:52.44 INFO  ==> Configuring replication parameters
postgresql-repmgr 09:35:52.48 INFO  ==> Configuring fsync
postgresql-repmgr 09:35:52.49 INFO  ==> Starting PostgreSQL in background...
postgresql-repmgr 09:35:52.50 INFO  ==> Creating repmgr user: repmgr
postgresql-repmgr 09:35:52.55 INFO  ==> Creating repmgr database: repmgr
postgresql-repmgr 09:35:52.90 INFO  ==> Stopping PostgreSQL...
waiting for server to shut down.... done
server stopped
postgresql-repmgr 09:35:53.11 INFO  ==> Starting PostgreSQL in background...
postgresql-repmgr 09:35:53.35 INFO  ==> Registering Primary...
postgresql-repmgr 09:35:53.74 INFO  ==> Loading custom scripts...
postgresql-repmgr 09:35:53.76 INFO  ==> Stopping PostgreSQL...
waiting for server to shut down.... done
server stopped
postgresql-repmgr 09:35:53.89 INFO  ==> ** PostgreSQL with Replication Manager setup finished! **

postgresql-repmgr 09:35:53.93 INFO  ==> Starting PostgreSQL in background...
waiting for server to start....2021-11-17 09:35:53.971 GMT [259] LOG:  pgaudit extension initialized
2021-11-17 09:35:53.972 GMT [259] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2021-11-17 09:35:53.972 GMT [259] LOG:  listening on IPv6 address "::", port 5432
2021-11-17 09:35:53.985 GMT [259] LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
2021-11-17 09:35:54.002 GMT [259] LOG:  redirecting log output to logging collector process
2021-11-17 09:35:54.002 GMT [259] HINT:  Future log output will appear in directory "/opt/bitnami/postgresql/logs".
2021-11-17 09:35:54.005 GMT [261] LOG:  database system was shut down at 2021-11-17 09:35:53 GMT
2021-11-17 09:35:54.012 GMT [259] LOG:  database system is ready to accept connections

Does this mean anything obvious to you?  Would it be useful for me to connect to the primary and see what "/opt/bitnami/postgresql/logs" says?

Thanks for any suggestions you can give.  I got a non-HA system up using Bitnami’s Charts with little effort but moving to an HA system is proving to be considerably more difficult ☹.

Paul DS

**Comment by juan131** [comment]
Hi @MswPaulDSmith

> Am I correct that the error is just one of documentation and you want me to update the Markdown that describes the currently undocumented setting?

No, Pablo was claiming that `global.pgpool.existingSecret` parameter does exist in the **values.yaml**, see:

- https://github.com/bitnami/charts/blob/master/bitnami/postgresql-ha/values.yaml#L45

While there's no `pgpool.existingSecret` parameter defined. See there's no "existingSecret" under "pgpool" object:

- https://github.com/bitnami/charts/blob/master/bitnami/postgresql-ha/values.yaml#L594

He was suggesting you to drop a PR to amend this.

> My postgresql/repmgr system is no longer complaining about passwords but still doesn’t come up.  The main issue seems to be this from what I assume is the back-up node…

Could you share the exact **values.yaml** you're using to install the chart at this point?

> Does this mean anything obvious to you?  Would it be useful for me to connect to the primary and see what "/opt/bitnami/postgresql/logs" says?

That's not necessary since the logs are already thrown to stdout. We're creating a symlink that connects the PostgreSQL log file with `/dev/stdout`, see:

- https://github.com/bitnami/bitnami-docker-postgresql-repmgr/blob/master/12/debian-10/rootfs/opt/bitnami/scripts/postgresql-repmgr/postunpack.sh#L25

**Comment by MswPaulDSmith** [comment]
Juan,

Thanks for your response.  Here is my values.yaml file (hapestfvmon is my Django-based app that sits above PostgreSQL-HA).  I've hacked out a couple of slightly sensitive bits of info using XXXX but the important information is all there.

I added the "pgpool.existingSecrets" section after receiving the e-mail from Pablo and it appeared to solve the password issue I was having at the time.  I noticed that the docs for the Chart didn't mention it but didn't appreciate that there is also a change needed to the default charts.

I'm pretty new to Kubernetes but am I correct that as written, my values.yaml would work because it explicitly defines pgpool.existingSecrets but if I had put this at global scope, it would not because the default values.yaml would not have caused it to be copied through because it is missing from the default definitions?

------------------------------------
# Note that memcache and postgresql are dependencies.
image: "docker-local.xxxxxxxx/cfs-pestfvmon"
imagePullPolicy: IfNotPresent
version: 0.1.0
hostname: "hapestfv.xxxxxxxx"
name: hapestfvmon
port: 8000
team: xxxxxxxx
jobs:
  - name: hapestfvmon-parsing
    schedule: "0 23 * * *"            # At 23:00 each night.
    command: /opt/pestfvmon/cron/pestfvmon_results.sh
    arguments: []
  - name: hapestfvmon-caching
    schedule: "0 7 * * *"             # At 07:00 each morning.
    command: python3
    args:
      - /opt/pestfvmon/cron/cache_pestfv.py
      - "--server"
      - hapestfvmon-service:8000

memcached:
  podLabels:
    app:
      hapestfvmon
  resources:
    requests:
      memory: 64Mi
    limits:
      memory: 64Mi

postgresql:
  primary:
    podLabels:
      app:
        hapestfvmon
  existingSecret: hapestfvmon-secrets
  postgresql:
    username: django
    database: pestfv
    repmgrUsername: repmgr
  persistence:
    mountPath: /var/lib/pestfvmon/postgresql-ha
  volumePermissions:
    enabled: true
  audit:
    logConnections: false
    logDisconnections: false
    logHostname: true
  readinessProbe:
    initialDelaySeconds: 60
  startupProbe:
    enabled: true
    initialDelaySeconds: 60

pgpool:
  primary:
    podLabels:
      app:
        hapestfvmon
  existingSecret: hapestfvmon-secrets
  customUsersSecret: hapestfvmon-secrets
-----------------------------


Paul DS

**Comment by juan131** [comment]
Hi @MswPaulDSmith 

>  Here is my values.yaml file 

Maybe you attached it to the mail but I'm unable to obtain it through GitHub UI. Could you please share it in a different way?

> am I correct that as written, my values.yaml would work because it explicitly defines pgpool.existingSecrets but if I had put this at global scope, it would not because the default values.yaml would not have caused it to be copied through because it is missing from the default definitions?

 Not really. AFAIK, you should be able to use `global.pgpool.existingSecrets` and it should work.

**Comment by MswPaulDSmith** [comment]
Juan,

I clicked on the "view it on GitHub" link below and I can see the values.yaml via the GitHub Issues system website.  You have to click through the "..." at the bottom of my e-mail as just the start is visible but if you click that, the values.yaml is there and visible.

Paul DS.

**Comment by juan131** [comment]
You're right! Sorry I missed it.

So, I guess that the secret `hapestfvmon-secrets` contains the following keys in the data:

- `postgresql-postgres-password`: including the password for "postgres" user.
- `postgresql-password`: including the password for "django" user.
- `repmgr-password`: including the password for "rempgr" user.
- `admin-password`: including the Pgpool admin user.
- `usernames`: including the custom usernames that will be added to `pgpool_passwd`.
- `passwords": including the custom passwords that will be added to `pgpool_passwd`.

Am I right?

Another question: why did you modify the default mount path? I mean, why did you set `persistence.mountPath` to  `/var/lib/pestfvmon/postgresql-ha`? Please note that the `bitnami/postgresql-repmgr` uses `/bitnami/postgresql` to store the data and it could be potentially problematic to switch the persistence mount point, see:

- https://github.com/bitnami/bitnami-docker-postgresql-repmgr#persisting-your-application

**Comment by MswPaulDSmith** [comment]
Juan,

You are correct about the secrets.  The persistent path was set link this just because my earlier no-HA system did this so that the data appeared on a sensible "standard Linux" type path but I spotted a couple of things I did wrong on the HA system and tried them - sadly with no success.

I had missed that the mountPath now appears to be outside of the postgresql section of values.yaml (which makes sense if repmgr also needs to access it) so I tried moving my definition, with no success.  I then commented out my definition so that I should have been using the bitnami default value of /bitnami/postgresql but this had not affect either.

  Note that I did stop the application and clean up all Kubernetes resources, including existing PVCs, between each attempt.

In both cases one of the Postgresql instances came up but the other just say in ContainerCreating state showing this problem (I've given you the full output in case it helps):

***@***.*** ~/git/pestfvmon$ kubectl describe pod hapestfvmon-postgresql-ha-postgresql-0
Name:           hapestfvmon-postgresql-ha-postgresql-0
Namespace:      fixed-voice-team-prod
Priority:       0
Node:           enfikubenode04/172.18.29.153
Start Time:     Mon, 29 Nov 2021 09:06:08 +0000
Labels:         app.kubernetes.io/component=postgresql
                app.kubernetes.io/instance=hapestfvmon
                app.kubernetes.io/managed-by=Helm
                app.kubernetes.io/name=postgresql-ha
                controller-revision-hash=hapestfvmon-postgresql-ha-postgresql-76468f67f9
                helm.sh/chart=postgresql-ha-7.11.1
                statefulset.kubernetes.io/pod-name=hapestfvmon-postgresql-ha-postgresql-0
Annotations:    kubernetes.io/psp: psp.baseline
                seccomp.security.alpha.kubernetes.io/pod: unconfined
Status:         Pending
IP:
IPs:            <none>
Controlled By:  StatefulSet/hapestfvmon-postgresql-ha-postgresql
Containers:
  postgresql:
    Container ID:
    Image:          docker.io/bitnami/postgresql-repmgr:11.13.0-debian-10-r57
    Image ID:
    Port:           5432/TCP
    Host Port:      0/TCP
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Liveness:       exec [bash -ec PGPASSWORD=$POSTGRES_PASSWORD psql -w -U "postgres" -d "postgres"  -h 127.0.0.1 -c "SELECT 1"] delay=30s timeout=5s period=10s #success=1 #failure=6
    Readiness:      exec [bash -ec PGPASSWORD=$POSTGRES_PASSWORD psql -w -U "postgres" -d "postgres"  -h 127.0.0.1 -c "SELECT 1"] delay=5s timeout=5s period=10s #success=1 #failure=6
    Environment:
      BITNAMI_DEBUG:                        false
      POSTGRESQL_VOLUME_DIR:                /bitnami/postgresql
      PGDATA:                               /bitnami/postgresql/data
      POSTGRES_USER:                        postgres
      POSTGRES_PASSWORD:                    <set to the key 'postgresql-password' in secret 'hapestfvmon-postgresql-ha-postgresql'>  Optional: false
      POSTGRES_DB:                          postgres
      POSTGRESQL_LOG_HOSTNAME:              true
      POSTGRESQL_LOG_CONNECTIONS:           false
      POSTGRESQL_LOG_DISCONNECTIONS:        false
      POSTGRESQL_PGAUDIT_LOG_CATALOG:       off
      POSTGRESQL_CLIENT_MIN_MESSAGES:       error
      POSTGRESQL_SHARED_PRELOAD_LIBRARIES:  pgaudit, repmgr
      POSTGRESQL_ENABLE_TLS:                no
      MY_POD_NAME:                          hapestfvmon-postgresql-ha-postgresql-0 (v1:metadata.name)
      REPMGR_UPGRADE_EXTENSION:             no
      REPMGR_PGHBA_TRUST_ALL:               no
      REPMGR_MOUNTED_CONF_DIR:              /bitnami/repmgr/conf
      REPMGR_NAMESPACE:                     fixed-voice-team-prod (v1:metadata.namespace)
      REPMGR_PARTNER_NODES:                 hapestfvmon-postgresql-ha-postgresql-0.hapestfvmon-postgresql-ha-postgresql-headless.$(REPMGR_NAMESPACE).svc.cluster.local,hapestfvmon-postgresql-ha-postgresql-1.hapestfvmon-postgresql-ha-postgresql-headless.$(REPMGR_NAMESPACE).svc.cluster.local,
      REPMGR_PRIMARY_HOST:                  hapestfvmon-postgresql-ha-postgresql-0.hapestfvmon-postgresql-ha-postgresql-headless.$(REPMGR_NAMESPACE).svc.cluster.local
      REPMGR_NODE_NAME:                     $(MY_POD_NAME)
      REPMGR_NODE_NETWORK_NAME:             $(MY_POD_NAME).hapestfvmon-postgresql-ha-postgresql-headless.$(REPMGR_NAMESPACE).svc.cluster.local
      REPMGR_LOG_LEVEL:                     NOTICE
      REPMGR_CONNECT_TIMEOUT:               5
      REPMGR_RECONNECT_ATTEMPTS:            3
      REPMGR_RECONNECT_INTERVAL:            5
      REPMGR_USERNAME:                      repmgr
      REPMGR_PASSWORD:                      <set to the key 'repmgr-password' in secret 'hapestfvmon-postgresql-ha-postgresql'>  Optional: false
      REPMGR_DATABASE:                      repmgr
    Mounts:
      /bitnami/postgresql from data (rw)
      /pre-stop.sh from hooks-scripts (rw,path="pre-stop.sh")
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-67thg (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  data-hapestfvmon-postgresql-ha-postgresql-0
    ReadOnly:   false
  hooks-scripts:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      hapestfvmon-postgresql-ha-postgresql-hooks-scripts
    Optional:  false
  default-token-67thg:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-67thg
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason                  Age                    From                     Message
  ----     ------                  ----                   ----                     -------
  Warning  FailedScheduling        6m20s                  default-scheduler        0/7 nodes are available: 7 pod has unbound immediate PersistentVolumeClaims.
  Warning  FailedScheduling        6m20s                  default-scheduler        0/7 nodes are available: 7 pod has unbound immediate PersistentVolumeClaims.
  Normal   Scheduled               6m12s                  default-scheduler        Successfully assigned fixed-voice-team-prod/hapestfvmon-postgresql-ha-postgresql-0 to enfikubenode04
  Normal   SuccessfulAttachVolume  6m13s                  attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-78d23434-603b-4063-8e0d-72e8d0cc5751"
  Warning  FailedMount             2m2s (x10 over 6m12s)  kubelet                  MountVolume.MountDevice failed for volume "pvc-78d23434-603b-4063-8e0d-72e8d0cc5751" : rpc error: code = NotFound desc = disk: 6000c29e7ccb2da150d74021cb48e2ee not attached to node
  Warning  FailedMount             115s (x2 over 4m10s)   kubelet                  Unable to attach or mount volumes: unmounted volumes=[data], unattached volumes=[hooks-scripts default-token-67thg data]: timed out waiting for the condition


Most importantly, you can see that the persistent data is now in the "standard bitnami location".

What should I be trying next to get to the bottom of this?  The fact that one instant of PostgreSQL seems happy seems really confusing to me - I would have expected either both to work, or both the fail!

Thanks,
Paul DS

**Comment by rafariossaa** [comment]
Hi,
  For the logs, it seems that the cluster is not able to assign a volume to the pod. Could you double check that point ?

**Comment by MswPaulDSmith** [comment]
Waiting on my Kubernetes IT guys to get a chance to check this; they've had a run of high sev issues to look at which has delayed this.

**Comment by rafariossaa** [comment]
No problem. Come back when you have news.

**Comment by diogomatsubara** [comment]
In the PR: https://github.com/bitnami/charts/pull/8797, I added the requested configuration `pgpool.existingSecrets` to `values.yaml` which fixes (at least for me!) the issue the OP reported. @pablogalegoc is that what you meant in your comment https://github.com/bitnami/charts/issues/8093#issuecomment-966354064?

**Comment by rafariossaa** [comment]
Hi,
  Thanks for the PR, we will review ti.
